ccl/changefeedccl: TestEncoders failed

ccl/changefeedccl.TestEncoders [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9880413?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9880413?buildTab=artifacts#/) on master @ [7d02e061759599a2b054ee06c71524443f77d20a](https://github.com/cockroachdb/cockroach/commits/7d02e061759599a2b054ee06c71524443f77d20a):


```
=== RUN   TestEncoders
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestEncoders1993582946
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestEncoders
    encoder_test.go:272: -- test log scope end --
--- FAIL: TestEncoders (0.02s)
=== RUN   TestEncoders/format=avro,envelope=wrapped,diff
    helpers_test.go:287: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:287
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/encoder_test.go:221
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/encoder_test.go:255
        	Error:      	Received unexpected error:
        	            	cannot decode binary record "foo_envelope" field "before": cannot decode binary union item 2: cannot decode binary record "foo_before" field "a": cannot decode binary union item 2: short buffer
        	Test:       	TestEncoders/format=avro,envelope=wrapped,diff
    --- FAIL: TestEncoders/format=avro,envelope=wrapped,diff (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEncoders.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27581

Epic CRDB-11732

ccl/changefeedccl.TestEncoders [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10122433?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10122433?buildTab=artifacts#/) on master @ [76078816b69ee6d1f8dacd581ef464660fb1b697](https://github.com/cockroachdb/cockroach/commits/76078816b69ee6d1f8dacd581ef464660fb1b697):


```
=== RUN   TestEncoders
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestEncoders3058303989
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestEncoders
    encoder_test.go:272: -- test log scope end --
--- FAIL: TestEncoders (0.05s)
=== RUN   TestEncoders/format=avro,envelope=key_only
    helpers_test.go:287: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:287
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/encoder_test.go:221
        	            				github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/encoder_test.go:255
        	Error:      	Received unexpected error:
        	            	cannot unmarshal schema JSON: unexpected end of JSON input
        	Test:       	TestEncoders/format=avro,envelope=key_only
    --- FAIL: TestEncoders/format=avro,envelope=key_only (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #102853 ccl/changefeedccl: TestEncoders failed [C-test-failure O-robot branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEncoders.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -1082,6 +1082,7 @@ func cdcTestNamedWithSystem(
 	t.Helper()
 	options := makeOptions(testOpts...)
 	cleanupCloudStorage := addCloudStorageOptions(t, &options)
+	TestingClearSchemaRegistrySingleton()
 
 	sinkType := randomSinkTypeWithOptions(options)
 	if sinkType == "skip" {

--- a/pkg/ccl/changefeedccl/schema_registry.go
+++ b/pkg/ccl/changefeedccl/schema_registry.go
@@ -373,3 +373,10 @@ type sharedSchemaRegistryCaches struct {
 }
 
 var schemaRegistrySingletons = &sharedSchemaRegistryCaches{cachePerEndpoint: make(map[string]*schemaRegistryCache)}
+
+// TestingClearSchemaRegistrySingleton clears out the singleton so that different tests don't pollute each other
+func TestingClearSchemaRegistrySingleton() {
+	schemaRegistrySingletons.mu.Lock()
+	defer schemaRegistrySingletons.mu.Unlock()
+	schemaRegistrySingletons.cachePerEndpoint = make(map[string]*schemaRegistryCache)
+}

